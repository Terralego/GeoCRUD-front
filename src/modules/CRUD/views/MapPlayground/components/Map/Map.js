import React, { useState, useContext, useEffect, useMemo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { useParams, useHistory } from 'react-router-dom';

import InteractiveMap, { INTERACTION_FN } from '@terralego/core/modules/Map/InteractiveMap';
import { CONTROL_CUSTOM, CONTROLS_TOP_LEFT } from '@terralego/core/modules/Map';

import { ACTION_CREATE, getLayers, getView } from '../../../../services/CRUD';
import { CRUDContext } from '../../../../services/CRUDProvider';
import { MapContext } from '../../../../services/MapProvider';

import { generateURI } from '../../../../config';
import Message from '../../../../../../components/Message';
import LayersControl from '../LayersControl';
import { usePrevious } from '../../../../../../utils/hooks';

const Map = ({ displayViewFeature, triggerFitBound }) => {
  const [interactions, setInteractions] = useState(null);
  const { feature, settings } = useContext(CRUDContext);

  const {
    addControl,
    controls,
    map,
    layers,
    loadSourceAndLayer,
    removeControl,
    setFitBounds,
    setMap,
    setInteractiveMapProps,
    sources,
  } = useContext(MapContext);

  const { id, layer } = useParams();

  const { push } = useHistory();

  const prevId = usePrevious(id);
  const prevFeature = usePrevious(feature[id]);

  const {
    i18n: {
      getResourceBundle,
      language,
      store: { options: { fallbackLng } },
    },
    t,
  } = useTranslation();

  const view = useMemo(() => getView(settings, layer), [layer, settings]);

  const backgroundStyle = useMemo(() => settings?.config?.BASE_LAYERS?.map(style => {
    const [label] = Object.keys(style);
    return {
      label,
      url: style[label].url,
    };
  }), [settings]);

  const mapConfig = useMemo(() => {
    if (!settings) {
      return null;
    }

    const config = settings?.config?.default.map;

    config.accessToken = config.mapbox_access_token;
    config.backgroundStyle = backgroundStyle;

    delete config.mapbox_access_token;

    return config;
  }, [backgroundStyle, settings]);

  const isFeatureID = useMemo(() => ![undefined, ACTION_CREATE].includes(id), [id]);

  const displayCurrentLayer = useCallback(() => {
    const {
      geometries = {},
    } = feature[id] || {};

    if (!map || !layers.length) {
      return;
    }

    const geometriesIdentifiers = Object.values(geometries)
      .map(({ identifier }) => identifier)
      .filter(Boolean);

    layers.forEach(layerItem => {
      if (!map.getLayer(layerItem.id)) return;
      const { main } = layerItem;
      const conditionalDisplay = layerItem.source === `${view.layer.id}` && (isFeatureID || main);
      map.setLayoutProperty(layerItem.id, 'visibility', conditionalDisplay ? 'visible' : 'none');
      if (geometriesIdentifiers.length && isFeatureID && !main) {
        map.setFilter(layerItem.id, ['in', '_id', ...geometriesIdentifiers]);
      }
    });
  }, [feature, id, isFeatureID, layers, map, view.layer.id]);

  useEffect(() => {
    if (!view || id) {
      setInteractions([]);
      return;
    }

    if (!displayViewFeature) {
      return;
    }

    const { layer: { id: layerId } = {} } = view;

    if (!layerId) {
      return;
    }

    const layersPaints = getLayers(settings);

    const nextInteractions = layersPaints
      .filter(({ source, main }) => source === `${layerId}` && main)
      .map(interaction => ({
        ...interaction,
        interaction: INTERACTION_FN,
        fn: ({
          feature: { sourceLayer, properties: { _id: propId } },
        }) => {
          push(generateURI('layer', { layer: sourceLayer, id: propId }));
        },
      }));

    setInteractions(nextInteractions);
  }, [displayViewFeature, id, push, settings, view]);

  const loadSourceAndLayerById = useCallback(
    layerId => {
      loadSourceAndLayer(settings, layerId);
    },
    [loadSourceAndLayer, settings],
  );

  useEffect(() => {
    if (!settings || !layer || !view || !map) {
      return;
    }

    const {
      layer: { id: layerId },
    } = view;

    loadSourceAndLayerById(layerId);

    displayCurrentLayer();
  }, [displayCurrentLayer, layer, loadSourceAndLayerById, map, settings, view]);

  const onFitBounds = useCallback(() => {
    if (!map || !view) {
      return;
    }

    if (!id && prevId) {
      return;
    }

    if (id === ACTION_CREATE) {
      return;
    }

    const { geom: { coordinates = [] } = {} } = feature[id] || {};
    const { extent: [w, s, e, n] } = view;

    setFitBounds({
      coordinates: isFeatureID ? coordinates : [[w, s], [e, n]],
      hasDetails: !!id,
    });

    // Exclude `prevId` from dependencies
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [feature, id, isFeatureID, map, setFitBounds, view]);

  useEffect(() => {
    onFitBounds();
  }, [onFitBounds, triggerFitBound]);


  const onMapLoaded = useCallback(nextMap => {
    setMap(nextMap);
  }, [setMap]);

  const onInit = useCallback(props => {
    setInteractiveMapProps(props);
  }, [setInteractiveMapProps]);

  const getMapStyle = useCallback(pk => getView(settings, pk, 'id').mapStyle, [settings]);

  useEffect(() => {
    if (!view) {
      return;
    }
    if (!id) {
      removeControl(CONTROL_CUSTOM);
      return;
    }

    if (prevFeature !== feature[id]) {
      removeControl(CONTROL_CUSTOM);
    }

    if (!isFeatureID) {
      const othersLayers = getLayers(settings).filter(
        ({ main, 'source-layer': sourceLayer }) => main && sourceLayer !== layer,
      );
      if (othersLayers.length) {
        addControl({
          control: CONTROL_CUSTOM,
          featureID: id,
          instance: LayersControl,
          layers: othersLayers,
          order: 1,
          position: CONTROLS_TOP_LEFT,
          loadSourceAndLayerById,
        });
      }
    } else {
      const { relations, geometries } = feature[id] || {};
      const layersProps = layers
        .filter(({ source, main }) => source === `${view.layer.id}` && !main)
        .map(item => ({
          ...item,
          empty: geometries?.[item['source-layer']].geom === null,
        }));

      if (layersProps.length && relations) {
        addControl({
          control: CONTROL_CUSTOM,
          featureID: id,
          getMapStyle,
          instance: LayersControl,
          layers: layersProps,
          order: 1,
          position: CONTROLS_TOP_LEFT,
          relations,
        });
      }
    }
  }, [
    addControl,
    feature,
    getMapStyle,
    id,
    isFeatureID,
    layer,
    layers,
    loadSourceAndLayerById,
    prevFeature,
    removeControl,
    settings,
    view,
  ]);

  const {
    terralego: { map: mapLocale },
  } = useMemo(() => (
    getResourceBundle(language.slice(0, 2)) || getResourceBundle(fallbackLng[0])
  ), [fallbackLng, getResourceBundle, language]);

  if ([mapConfig, interactions].includes(null)) {
    return null;
  }

  if (!backgroundStyle.length) {
    return (
      <Message intent="danger" className="CRUD-no-map">
        {t('CRUD.settings.unableToLoadMap')}
      </Message>
    );
  }

  return (
    <InteractiveMap
      {...mapConfig}
      controls={controls}
      customStyle={{ layers, sources }}
      interactions={interactions}
      onInit={onInit}
      onMapLoaded={onMapLoaded}
      onStyleChange={displayCurrentLayer}
      locale={mapLocale}
      translate={t}
    />
  );
};

Map.propTypes = {
  displayViewFeature: PropTypes.bool,
  triggerFitBound: PropTypes.string,
};

Map.defaultProps = {
  displayViewFeature: true,
  triggerFitBound: undefined,
};

export default Map;
