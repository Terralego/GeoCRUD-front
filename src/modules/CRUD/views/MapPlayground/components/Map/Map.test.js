import React from 'react';
import Router from 'react-router-dom';
import renderer from 'react-test-renderer';
import Map from './Map';
import { CRUDContext } from '../../../../services/CRUDProvider';
import { MapContext } from '../../../../services/MapProvider';

jest.mock('@terralego/core/modules/Map', () => ({
  CONTROL_CUSTOM: 'custom-control',
  CONTROLS_TOP_LEFT: 'control-top-left',
}));

jest.mock('@terralego/core/modules/Map/InteractiveMap', () => {
  const InteractiveMap = props => <div {...props} />;
  InteractiveMap.INTERACTION_FN = 'function';
  return InteractiveMap;
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    id: undefined,
    layer: undefined,
  }),
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    i18n: {
      getResourceBundle: jest.fn(() => ({
        terralego: {
          map: 'mapLocale',
        },
      })),
      language: 'en-en',
      store: {
        options: { fallbackLng: 'en' },
      },
    },
    t: text => text,
  }),
}));

jest.mock('../LayersControl', () => props => <div {...props} />);

const defaultCrudContextProps = {
  feature: { fooId: {} },
  settings: {
    menu: [
      {
        crud_views: [
          {
            extent: [1, 2, 3, 4],
            layer: {
              id: 'fooId',
              name: 'fooLayer',
            },
            map_layers: [
              {
                id_layer_vt: 'fooId',
                name: 'fooLayer',
                style: {},
                main: true,
              },
            ],
          },
          {
            extent: [1, 2, 3, 4],
            layer: {
              id: 'barId',
              name: 'barLayer',
            },
            map_layers: [
              {
                id_layer_vt: 'barId',
                name: 'barLayer',
                style: {},
              },
            ],
          },
        ],
      },
    ],
    config: {
      BASE_LAYERS: [
        { OSM: { id: '1', url: 'endpointToTilejson' } },
        { Mapbox: { id: '2', url: 'endpointToTilejson' } },
      ],
      default: {
        map: {
          mapbox_access_token: 'FAKE_ACCESS_TOKEN',
        },
      },
    },
  },
};

const defaultMapContextProps = {
  addControl: jest.fn(),
  controls: [
    {
      control: 'NavigationControl',
      position: 'top-right',
    },
  ],
  map: {
    getLayer: jest.fn(),
    setFilter: jest.fn(),
    setLayoutProperty: jest.fn(),
  },
  layers: [
    {
      id: 'fooId',
      'source-layer': 'fooLayer',
      source: 'my-source-id',
      type: 'circle',
      paint: {
        'circle-color': '#0075d1',
        'circle-radius': 6,
      },
      title: 'Foo',
      main: true,
      weight: 850,
    },
    {
      id: 'barId',
      'source-layer': 'fooLayer',
      source: 'my-source-id',
      type: 'fill',
      paint: {
        'fill-color': 'black',
        'fill-opacity': 0.4,
      },
      title: 'Bar',
      main: false,
      weight: 850,
    },
  ],
  loadSourceAndLayer: jest.fn(),
  removeControl: jest.fn(),
  setFitBounds: jest.fn(),
  setMap: jest.fn(),
  setInteractiveMapProps: jest.fn(),
  sources: [
    {
      id: 'my-source-id',
      type: 'vector',
      url: '/api/crud/layers/my-source-id/tilejson/',
    },
  ],
};

const MapWithProviders = ({ crudContextProps = {}, mapContextProps = {}, ...props }) => (
  <CRUDContext.Provider value={{ ...defaultCrudContextProps, ...crudContextProps }}>
    <MapContext.Provider value={{ ...defaultMapContextProps, ...mapContextProps }}>
      <Map {...props} />
    </MapContext.Provider>
  </CRUDContext.Provider>
);

describe('should render correctly', () => {
  it('map without object', () => {
    const tree = renderer.create(<MapWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('map focus on layer', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ layer: 'fooLayer' });
    const tree = renderer.create(<MapWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('map focus on geom object', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ layer: 'fooLayer', id: 'fooId' });
    const tree = renderer.create(<MapWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
