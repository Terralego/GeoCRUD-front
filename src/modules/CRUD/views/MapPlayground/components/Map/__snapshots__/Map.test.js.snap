// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly map focus on geom object 1`] = `
<div
  backgroundStyle={
    Array [
      Object {
        "label": "OSM",
        "url": "endpointToTilejson",
      },
      Object {
        "label": "Mapbox",
        "url": "endpointToTilejson",
      },
    ]
  }
  controls={
    Array [
      Object {
        "control": "NavigationControl",
        "position": "top-right",
      },
    ]
  }
  customStyle={
    Object {
      "layers": Array [
        Object {
          "id": "fooId",
          "main": true,
          "paint": Object {
            "circle-color": "#0075d1",
            "circle-radius": 6,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Foo",
          "type": "circle",
          "weight": 850,
        },
        Object {
          "id": "barId",
          "main": false,
          "paint": Object {
            "fill-color": "black",
            "fill-opacity": 0.4,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Bar",
          "type": "fill",
          "weight": 850,
        },
      ],
      "sources": Array [
        Object {
          "id": "my-source-id",
          "type": "vector",
          "url": "/api/crud/layers/my-source-id/tilejson/",
        },
      ],
    }
  }
  interactions={Array []}
  locale="mapLocale"
  onInit={[Function]}
  onMapLoaded={[Function]}
  onStyleChange={[Function]}
  translate={[Function]}
/>
`;

exports[`should render correctly map focus on layer 1`] = `
<div
  backgroundStyle={
    Array [
      Object {
        "label": "OSM",
        "url": "endpointToTilejson",
      },
      Object {
        "label": "Mapbox",
        "url": "endpointToTilejson",
      },
    ]
  }
  controls={
    Array [
      Object {
        "control": "NavigationControl",
        "position": "top-right",
      },
    ]
  }
  customStyle={
    Object {
      "layers": Array [
        Object {
          "id": "fooId",
          "main": true,
          "paint": Object {
            "circle-color": "#0075d1",
            "circle-radius": 6,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Foo",
          "type": "circle",
          "weight": 850,
        },
        Object {
          "id": "barId",
          "main": false,
          "paint": Object {
            "fill-color": "black",
            "fill-opacity": 0.4,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Bar",
          "type": "fill",
          "weight": 850,
        },
      ],
      "sources": Array [
        Object {
          "id": "my-source-id",
          "type": "vector",
          "url": "/api/crud/layers/my-source-id/tilejson/",
        },
      ],
    }
  }
  interactions={
    Array [
      Object {
        "fn": [Function],
        "id": "CRUD-fooId-fooId",
        "interaction": "function",
        "main": true,
        "name": "fooLayer",
        "source": "fooId",
        "source-layer": "fooId",
      },
    ]
  }
  locale="mapLocale"
  onInit={[Function]}
  onMapLoaded={[Function]}
  onStyleChange={[Function]}
  translate={[Function]}
/>
`;

exports[`should render correctly map without object 1`] = `
<div
  accessToken="FAKE_ACCESS_TOKEN"
  backgroundStyle={
    Array [
      Object {
        "label": "OSM",
        "url": "endpointToTilejson",
      },
      Object {
        "label": "Mapbox",
        "url": "endpointToTilejson",
      },
    ]
  }
  controls={
    Array [
      Object {
        "control": "NavigationControl",
        "position": "top-right",
      },
    ]
  }
  customStyle={
    Object {
      "layers": Array [
        Object {
          "id": "fooId",
          "main": true,
          "paint": Object {
            "circle-color": "#0075d1",
            "circle-radius": 6,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Foo",
          "type": "circle",
          "weight": 850,
        },
        Object {
          "id": "barId",
          "main": false,
          "paint": Object {
            "fill-color": "black",
            "fill-opacity": 0.4,
          },
          "source": "my-source-id",
          "source-layer": "fooLayer",
          "title": "Bar",
          "type": "fill",
          "weight": 850,
        },
      ],
      "sources": Array [
        Object {
          "id": "my-source-id",
          "type": "vector",
          "url": "/api/crud/layers/my-source-id/tilejson/",
        },
      ],
    }
  }
  interactions={Array []}
  locale="mapLocale"
  onInit={[Function]}
  onMapLoaded={[Function]}
  onStyleChange={[Function]}
  translate={[Function]}
/>
`;
